src/ts-node.ts

```diff
@@ -1,16 +1,21 @@
 import * as fs from 'fs'
 import * as path from 'path'
 import * as TSNode from 'ts-node'
-import {parseConfigFileTextToJson} from 'typescript'
 
 import Debug from './debug'
+const debug = Debug()
+
+let typescript: typeof import('typescript')
```

Comment:
You sir, are thinking like JavaScript! This was my original implementation, but TypeScript got very cranky with me since the type for `let typescript` could not be inferred automagically.

Reply:
Thank you for the explanation! 

I am still not Typescript convinced, and I come from C# and Java background. I just haven't come across issues that I can see typescript solving that proper testing, prs, lint, and other process can't solve. I also can't see typescript solving the problems of those other areas, so they would still be required.

Reply:
I am a JavaScript person as well, but hey – open source means exploring new things 🤷‍♂️

Reply:
I'm not entirely sure myself why this works (but it does, the type is correct here and not a promise).

I think that `import()` behaves differently in the value position vs the type position.

value position (promise):

```typescript
async function () {
  let typescript = await import('typescript')
}
```

type position:

```typescript
async function () {
  let typescript: typeof import('typescript') = await import('typescript')
}
```

dynamic imports are admittedly not very clean in typescript and this is weird behavior in any case

Reply:
<img width="1082" alt="screen shot 2019-01-31 at 10 45 33" src="https://user-images.githubusercontent.com/216188/52077236-5cb75280-2545-11e9-8163-892c99053a28.png">

Reply:
So what's the type if typescript isn't installed and this code runs?

```
let typescript: typeof import('typescript')```

Reply:
there are no types when the code is running, typescript types are only used at compile-time

Reply:
So I guess it's working because since typescript has to run to compile this source code, typescript is available at compile time in this scenario. So it just works. I wonder if this would work if the dependency was something other than typescript.

PS. Sorry for the chatty conversation, I fully respect your inboxes!

Reply:
@blackdynamo - I agree with you. While it works, I think it's technically incorrect that `typescript` is being required at runtime, but not required as a dependency.

https://github.com/oclif/config/pull/72#discussion_r252495685

Reply:
Okay, storytime. Back in the day there was a dream – a dream called [`optionalDependencies`](https://docs.npmjs.com/files/package.json#optionaldependencies). 

The current behavior of `optionalDependencies` is just a tiny fraction of what was discussed when it was being planned ... at least 6 years ago.

Behavior that would have covered this scenario: **opt-in dependencies**. What I mean by that is `typescript` is not _actually_ a run-time dependency. It is a "nice to have" that only attempts to parse `tsconfig.json` file when present. 

There are a lot of grey areas like this in production scenarios for package management. Lazy require is a well known technique to handle them which imho is appropriate here.

Reply:
TIL! Thanks for the further explanation 👍